accepted/2021/InboxSourceGenerators.md

```diff
@@ -0,0 +1,165 @@
+
+So for means of detection we have:
+1. Manual
+2. PDB indicator: eg source generator assembly version, file version, hash)
+3. Assembly metadata in consuming user project: eg AssemblyMetadata("SourceGeneratedBy", "source generator assembly version, file version, hash")
+4. Runtime breadcrumbs
+
+For means of update we have:
+1. No notification, user needs to subscribe to servicing updates.
+2. Design time notification.
+3. Killbit in the runtime to prevent execution of user assemblies that contain vulnerable source generated code.
+4. ~~Automatically update user assembly though a hotpatch applied by runtime or JIT~~
+
+We'll need to better understand what customer's expectations are around servicing to decide if we should do any work on these axes.  My initial inclination is to only have a PDB indicator and require users to subscribe to updates.
+
+### Functionality concerns for source generators
+
+Though the following are not specific to inbox source generators, these concerns are worth reviewing to describe how they apply to framework specific source generators.
+
+Source generator assemblies should target .NETStandard and must run on both .NETFramework and .NETCore. Source generators should support generating source for all TargetFrameworks supported by their companion library.  In the case of inbox source generators discussed in this document this need only be the TargetFramework the generator ships in.  In the case of a NuGet package provided source generator, this should be all frameworks supported by package.
+
+Source generators should limit their dependencies.  There is no scheme for encoding framework-specific nor runtime-specific implementations of dependencies or source generators based on the runtime environment so source generators should avoid any dependencies on assemblies which need to differ by framework or runtime.  Source generators may depend on packages which are known dependencies of the compiler and should depend on versions less than or equal to that provided by the compiler and should not include that dependency in their package.
+
+Source generators should not directly reference nor execute runtime code.  Despite inbox source generators being coupled to a specific framework version, source generators cannot assume they will be running on that framework.  They may be running on a newer, older, or completely different framework.  There is no type-equivalence guaranteed between references of the source generator and those of the user code that is under analysis.  As such source generators will need to examine user code references by name, or when needing to make equivalence checks can resolve types in the user code's type reference set to ensure valid equivalence checks.  Source sharing should be considered as an alternative for cases where source generators need to execute runtime library code.
```

Comment:
> Source generators should not directly reference nor execute runtime code.

Maybe it's because the coffee hasn't kicked in yet, but I'm at a loss as to what this means.  The source generators are .NET assemblies, intended to make some algorithmic decisions, and act on them... seems like things like GetEnumerator() and string.Format are going to be highly utilized.

Maybe it's saying that we shouldn't be providing public API in the 'target' library to be a "generator helper", where the output of a given version of the generator could vary by the runtime version used by a particular compiler invocation.  (I can't come up with any phrasing of this concept that would enable us to write a generator for (or to avoid) Regex without using a weasel phrase like "generator helper", since the Regex source generator could potentially use Regex as a "stable platform component" as an implementation detail)

Reply:
This is meant to remind people that they shouldn't treat references as executable assemblies.  We have a lot of runtime developers that we're asking to write design time code and they need to understand the difference and best practice.

>  since the Regex source generator could potentially use Regex as a "stable platform component" as an implementation detail

Be very careful about that.  You'd need to be sure that the Regex in .NETFramework is sufficient for those purposes.  You'd also need to bridge between the developer's source references and the framework component.

I will try to clarify this.